WIP✨(backend) refactor indexation pipeline

joehybird opened the pull request:
Add support for deferred loading, preprocessing & embedding of documents.

## TODO

- [ ] Throttle indexation tasks or create commands + cron
- [x] Parser for pdf files with albert
- [ ] Download directly from storage APIs (S3)
- [ ] Use base64 + "encoding=base64" argument for small binary files sent directly through `index/` endpoint

## Whishlist

- [ ] Add async support for download (use asyncio loop)
- [ ] Parser for all formats with a dedicated service (docling, unstructured, ...)
- [ ] Use hash to prevent indexing the same content multiple times

joehybird committed:
✨(backend) refactor indexation pipeline

Add support for deferred loading, preprocessing & embedding of documents.
Add mimetype, language, content_status & content_uri fields in document schema.

Signed-off-by: Fabre Florian <ffabre@hybird.org>

qbey commented on src/backend/core/schemas.py:
I don't understand how are we supposed to pass a `content_uri` via the API?

joehybird replied:
We'll need bit of documentation here :grin: The idea is to give an url to the API instead of the content when the file is too large in order to download it from Find later.
It can be an url to an endpoint of the service or, a better idea (thx @lunika), a direct link to the storage

mascarpon3 commented on src/backend/core/models.py:
```diff
@@ -73,3 +80,59 @@ def generate_secure_token():
+    @property
+    def is_loaded(self):
+        """Retuns true if in loaded status"""
+        return self.content_status == enums.ContentStatusEnum.LOADED
```
those properties are so trivial I would rather rely on `content_status` only. 

Also isn't "`indexing_status`" more descriptive than "`content_status`" ?

mascarpon3 commented on src/backend/core/schemas.py:
```diff
@@ -38,6 +38,8 @@ class DocumentSchema(BaseModel):
     reach: Optional[enums.ReachEnum] = Field(default=enums.ReachEnum.RESTRICTED)
     is_active: bool
+    mimetype: Annotated[str, Field(default="text/plain")]
+    language: Annotated[str, Field(default="en")]
```
why do you need to introduce a `language` field here ? 

I am handling multi-language in my pr about full-text search so I recommand you remove it if you are not using it.

joehybird committed:
✨(backend) albert AI client & pdf conversion

Add AlbertAI client to wrap embedding & conversion API calls
Implement working pdf to markdown converter using Albert

Signed-off-by: Fabre Florian <ffabre@hybird.org>

joehybird committed:
✨(backend) fix service index prefixed name in create_demo

Use service.index_name instead of service.name in create_demo
command.

Signed-off-by: Fabre Florian <ffabre@hybird.org>

joehybird committed:
✨(backend) add preprocess support for indexer

New processors mechanism in IndexerTaskService : after the loading & conversion
steps a list a functions can be chained to transform the document content (like
django middlewares)

Signed-off-by: Fabre Florian <ffabre@hybird.org>

StephanMeijer commented:
Thank you dearly for your contribution. I noticed this PR stalled and it will be revised in the coming version(s) anyhow, therefore closing.
